pkg/rancher-desktop/main/extensions/manager.ts

```diff
@@ -267,6 +267,9 @@ export class ExtensionManagerImpl implements ExtensionManager {
       })(repo, tag));
     }
     await Promise.all(tasks);
+
+    // Register a listener to shut down extensions on quit
+    mainEvents.on('quit', this.triggerExtensionShutdown);
```

Comment:
It is not clear to me how the application makes sure all the extensions can finish their shutdown before the application exits.

Reply:
We don't?

Reply:
So there is no guarantee that the extension will be able to stop Ollama before the app quits, so factory reset may still fail? Doesn't this kind of defeat the purpose of having the hook?

Reply:
No, factory reset _still_ wouldn't run this at all, because we `kill -9 rancher-desktop` and don't attempt to gracefully shut down.  We will have a separate code path to terminate any processes whose executable is from inside the extension directory instead.

Reply:
Ok, my bad. This is just supposed to stop Ollama when it has been started by the extension. So shouldn't we still wait for it to finish shutting down? I don't understand why we would want this hook when could exit before it actually executes. The same way we wait for the backend to terminate before we (gracefully) exit.

It is totally possible that I'm still just confused, but I don't yet get it. Or do we run the hooks before stopping the backend, and just expect the hooks to be done before the VM has stopped?

Reply:
We run the hooks before stopping the backend, but I thought it would be bad to wait for it to finish (because it can take arbitrarily long, stopping the application from exiting).  Currently, we notify the extensions that we'll exit, and their process can do whatever it likes.

If you think it would make more sense to block shutdown on the extensions, that's doable too.

Reply:
> their process can do whatever it likes.

Thank you, it was this part that I was missing: they are running in their own process and not in some async function in the app itself.

Are they processes on the host, or running inside a container in the guest? If they are host processes, then I think it is all good if the notification are sent before the backend shuts down.

If they are running inside a container, then there may still be a race.

Reply:
They are processes on the host, as in `something.exe`.

But yes, we may need to at least make sure they all get _started_ before quitting for real; currently that might not happen consistently.

Reply:
Just for illustration, I have recently written this code, and had to insert the `sleep` to be able to still send out the response to the client before shutting down the server:

```go
	router.PUT("/v1/shutdown", func(c *gin.Context) {
		go func() {
			// Allow the response to be sent back before shutting down
			time.Sleep(1 * time.Second)
			_ = syscall.Kill(syscall.Getpid(), syscall.SIGTERM)
		}()
		c.JSON(200, gin.H{"status": "Server is shutting down"})
	})
```

So I am/was worried we are in a similar situation here

Reply:
And now I cannot remember why I didn't simply sent the kill signal after sending the response? I have to retry that to see if there was a reason that didn't work.

Reply:
idk, simplistic testing shows that it works just fine without the goroutine, so no idea how I ended up with that contortion before.